File: src/services/tauri/filesystem.ts
import {
  getJson,
  invokeTauri,
  isDesktopTauri,
  postJson,
  postJsonWithOptions,
  postNoContent,
  postNoContentWithOptions,
} from './shared';

export async function readFile(filePath: string, options?: { suppressToast?: boolean }) {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('read_file', { filePath });
  }
  return await postJsonWithOptions<string>(
    '/api/filesystem/read-file',
    { filePath },
    options
  );
}

export async function readTextFileLines(filePath: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('read_text_file_lines', { filePath });
  }
  return await postJson<string[]>('/api/filesystem/read-text-file-lines', { filePath });
}

export async function getCodexHome() {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('codex_home');
  }
  return await getJson<string>('/api/filesystem/codex-home');
}

export async function writeFile(filePath: string, content: string) {
  if (isDesktopTauri()) {
    await invokeTauri('write_file', { filePath, content });
    return;
  }
  await postNoContent('/api/filesystem/write-file', { filePath, content });
}

export async function readPdfContent(filePath: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('read_pdf_content', { filePath });
  }
  return await postJson<string>('/api/filesystem/read-pdf', { filePath });
}

export async function readXlsxContent(filePath: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('read_xlsx_content', { filePath });
  }
  return await postJson<string>('/api/filesystem/read-xlsx', { filePath });
}

export async function readDirectory(path: string, options?: { suppressToast?: boolean }) {
  if (isDesktopTauri()) {
    return await invokeTauri<Array<import('./shared').TauriFileEntry>>('read_directory', { path });
  }
  return await postJsonWithOptions<Array<import('./shared').TauriFileEntry>>(
    '/api/filesystem/read-directory',
    { path },
    options
  );
}

export async function getHomeDirectory() {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('get_home_directory');
  }
  return await getJson<string>('/api/filesystem/home-directory');
}

export async function searchFiles(params: {
  root: string;
  query: string;
  excludeFolders: string[];
  maxResults?: number;
}) {
  if (isDesktopTauri()) {
    return await invokeTauri<Array<import('./shared').TauriFileEntry>>('search_files', params);
  }
  return await postJson<Array<import('./shared').TauriFileEntry>>('/api/filesystem/search-files', {
    root: params.root,
    query: params.query,
    exclude_folders: params.excludeFolders,
    max_results: params.maxResults,
  });
}

export async function searchFilesByName(params: {
  root: string;
  query: string;
  excludeFolders: string[];
  maxResults?: number;
}) {
  if (isDesktopTauri()) {
    return await invokeTauri<Array<import('./shared').TauriFileEntry>>('search_files_by_name', params);
  }
  return await postJson<Array<import('./shared').TauriFileEntry>>('/api/filesystem/search-files-by-name', {
    root: params.root,
    query: params.query,
    exclude_folders: params.excludeFolders,
    max_results: params.maxResults,
  });
}

export async function canonicalizePath(path: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('canonicalize_path', { path });
  }
  return await postJson<string>('/api/filesystem/canonicalize-path', { path });
}

export async function deleteFile(filePath: string) {
  if (isDesktopTauri()) {
    await invokeTauri('delete_file', { filePath });
    return;
  }
  await postNoContent('/api/filesystem/delete-file', { filePath });
}

export async function startWatchDirectory(folderPath: string) {
  if (isDesktopTauri()) {
    await invokeTauri('start_watch_directory', { folderPath });
    return;
  }
  await postNoContent('/api/filesystem/start-watch', { path: folderPath });
}

export async function stopWatchDirectory(folderPath: string) {
  if (isDesktopTauri()) {
    await invokeTauri('stop_watch_directory', { folderPath });
    return;
  }
  await postNoContent('/api/filesystem/stop-watch', { path: folderPath });
}

export async function startWatchFile(filePath: string) {
  if (isDesktopTauri()) {
    await invokeTauri('start_watch_file', { filePath });
    return;
  }
  await postNoContentWithOptions('/api/filesystem/start-watch-file', { filePath }, { suppressToast: true });
}

export async function stopWatchFile(filePath: string) {
  if (isDesktopTauri()) {
    await invokeTauri('stop_watch_file', { filePath });
    return;
  }
  await postNoContentWithOptions('/api/filesystem/stop-watch-file', { filePath }, { suppressToast: true });
}

File: src-tauri/src/commands/filesystem.rs
use crate::features::filesystem::{
    directory_ops, file_io,
    file_parsers::{pdf, xlsx},
    file_types::FileEntry,
    watch,
};
use crate::state::WatchState;
use std::sync::Arc;
use tauri::{AppHandle, Emitter, State};

#[tauri::command]
pub async fn read_directory(path: String) -> Result<Vec<FileEntry>, String> {
    directory_ops::read_directory(path).await
}

#[tauri::command]
pub async fn get_home_directory() -> Result<String, String> {
    directory_ops::get_home_directory().await
}

#[tauri::command]
pub async fn search_files(
    root: String,
    query: String,
    exclude_folders: Vec<String>,
    max_results: Option<usize>,
) -> Result<Vec<FileEntry>, String> {
    directory_ops::search_files(root, query, exclude_folders, max_results).await
}

#[tauri::command]
pub async fn canonicalize_path(path: String) -> Result<String, String> {
    directory_ops::canonicalize_path(path).await
}

#[tauri::command]
pub async fn read_file(file_path: String) -> Result<String, String> {
    file_io::read_file(file_path).await
}

#[tauri::command]
pub async fn read_text_file_lines(file_path: String) -> Result<Vec<String>, String> {
    file_io::read_text_file_lines(file_path).await
}

#[tauri::command]
pub async fn write_file(file_path: String, content: String) -> Result<(), String> {
    file_io::write_file(file_path, content).await
}

#[tauri::command]
pub async fn delete_file(file_path: String) -> Result<(), String> {
    file_io::delete_file(file_path).await
}

#[tauri::command]
pub async fn read_pdf_content(file_path: String) -> Result<String, String> {
    pdf::read_pdf_content(file_path).await
}

#[tauri::command]
pub async fn read_xlsx_content(file_path: String) -> Result<String, String> {
    xlsx::read_xlsx_content(file_path).await
}

fn tauri_watch_emitter(app: AppHandle) -> Arc<dyn Fn(watch::FsChangePayload) + Send + Sync> {
    Arc::new(move |payload: watch::FsChangePayload| {
        let _ = app.emit("fs_change", &payload);
    })
}

#[tauri::command]
pub async fn start_watch_directory(
    app: AppHandle,
    state: State<'_, WatchState>,
    folder_path: String,
) -> Result<(), String> {
    watch::start_watch_directory(state.inner(), folder_path, tauri_watch_emitter(app)).await
}

#[tauri::command]
pub async fn stop_watch_directory(
    state: State<'_, WatchState>,
    folder_path: String,
) -> Result<(), String> {
    watch::stop_watch_directory(state.inner(), folder_path).await
}

#[tauri::command]
pub async fn start_watch_file(
    app: AppHandle,
    state: State<'_, WatchState>,
    file_path: String,
) -> Result<(), String> {
    watch::start_watch_file(state.inner(), file_path, tauri_watch_emitter(app)).await
}

#[tauri::command]
pub async fn stop_watch_file(
    state: State<'_, WatchState>,
    file_path: String,
) -> Result<(), String> {
    watch::stop_watch_file(state.inner(), file_path).await
}
